File: haskell/palindrome.hs
lowerChars = ['a'..'z']
upperChars = ['A'..'Z']
alphaChars = lowerChars ++ upperChars
charPairs = zip lowerChars upperChars

toLower :: Char -> Char
toLower inputChar
  | inputChar `elem` upperChars = head [a | (a, b) <- charPairs, inputChar == b]
  | otherwise = inputChar

cleanString :: [Char] -> [Char]
cleanString "" = ""
cleanString inputString = [toLower x | x <- inputString, x `elem` alphaChars]

--reverse and check for equality
palindrome :: [Char] -> Bool
palindrome "" = False
palindrome inputString = (cleanedString == reverse cleanedString)
  where cleanedString = cleanString inputString

File: javascript/palindrome.js
var alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
var upperChars = alphabet.split( "" );
var lowerChars = alphabet.toLowerCase().split( "" );
var alphaChars = lowerChars.concat( upperChars );

function isAlpha( inputChar ) {
  return alphaChars.includes( inputChar );
}

function toLowerCase( inputString ) {
  return inputString.toLowerCase();
}

function cleanString( inputString ){
  return inputString.split( "" )
                    .filter( isAlpha )
                    .map( toLowerCase )
                    .join("");
}

function stringReverse( inputString ) {
  return inputString.split( "" )
                    .reverse()
                    .join( "" );
}

function palindrome( inputString ) {
  var cleanedString = cleanString( inputString );
  return cleanedString === stringReverse( cleanedString );
}
